<a href="https://colab.research.google.com/github/fundamentals-of-data-science/course-materials/blob/master/labs/Lab_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 2 Lab - Probability

## General Instructions

In this course, Labs are the chance to applying concepts and methods discussed in the module.
They are a low stakes (pass/fail) opportunity for you to try your hand at *doing*.
Please make sure you follow the general Lab instructions, described in the Syllabus.
The summary is:

* Discussions should start as students work through the material, first Wednesday at the start of the new Module week. 
* Labs are due by **Sunday**.
* Lab solutions are released Monday.  
* Post Self Evaluation and Lab to Lab Group on Blackboard and Lab to Module on Blackboard on **Monday**.

The last part is important because the Problem Sets will require you to perform the same or similar tasks without guidance.
Problem Sets are your opportunity to demonstrate that you understand how to apply the concepts and methods discussed in the relevant Modules and Labs.

## Specific Instructions

1.  For Blackboard submissions, if there are no accompanying files, you should submit *only* your notebook and it should be named using *only* your JHED id: fsmith79.ipynb for example if your JHED id were "fsmith79". If the assignment requires additional files, you should name the *folder/directory* your JHED id and put all items in that folder/directory, ZIP it up (only ZIP...no other compression), and submit it to Blackboard.

    * do **not** use absolute paths in your notebooks. All resources should located in the same directory as the rest of your assignments.
    * the directory **must** be named your JHED id and **only** your JHED id.
    * do **not** return files provided by us (data files, .py files)


2. Data Science is as much about what you write (communicating) as the code you execute (researching). In many places, you will be required to execute code and discuss both the purpose and the result. Additionally, Data Science is about reproducibility and transparency. This includes good communication with your team and possibly with yourself. Therefore, you must show **all** work.

3. Avail yourself of the Markdown/Codecell nature of the notebook. If you don't know about Markdown, look it up. Your notebooks should not look like ransom notes. Don't make everything bold. Clearly indicate what question you are answering.

4. Submit a cleanly executed notebook. The first code cell should say `In [1]` and each successive code cell should increase by 1 throughout the notebook.

In [15]:
from pprint import pprint

## Manipulating and Interpreting Probability

Given the following *joint probability distribution*, $P(A,B)$, for $A$ and $B$,

```
|    | a1   | a2   |
|----|------|------|
| b1 | 0.37 | 0.16 |
| b2 | 0.23 | ?    |
```

Answer the following questions.

**1\. What is $P(A=a2, B=b2)$?**

$P(A=a2, B=b2) = 1 - (0.37 + 0.16 + 0.23) = 0.24$

**2\. If I observe events from this probability distribution, what is the probability of seeing (a1, b1) then (a2, b2)?**

$P(a1, b1) \times P(a2, b2) = 0.37 \times 0.24 = 0.0888$

**3\. Calculate the marginal probability distribution, $P(A)$.**

$P(A=a1) = 0.37 + 0.23 = 0.60$

$P(A=a2) = 0.16 + 0.24 = 0.40$

**4\. Calculate the marginal probability distribution, $P(B)$.**

$P(B=b1) = 0.37 + 0.16 = 0.53$

$P(B=b2) = 0.23 + 0.24 = 0.47$

**5\. Calculate the conditional probability distribution, $P(A|B)$.**

$P(A=a1|B=b1) = \dfrac{0.37}{0.53} \approx 0.698$

$P(A=a2|B=b1) = \dfrac{0.16}{0.53} \approx 0.302$

$P(A=a1|B=b2) = \dfrac{0.23}{0.47} \approx 0.489$

$P(A=a2|B=b2) = \dfrac{0.24}{0.47} \approx 0.511$

**6\. Calculate the conditional probability distribution, $P(B|A)$.**

$P(B=b1|A=a1) = \dfrac{0.37}{0.60} \approx 0.617$

$P(B=b2|A=a1) = \dfrac{0.23}{0.60} \approx 0.383$

$P(B=b1|A=a2) = \dfrac{0.16}{0.40} = 0.400$

$P(B=b2|A=a2) = \dfrac{0.24}{0.40} = 0.600$

**7\. Does $P(A|B) = P(B|A)$? What do we call the belief that these are always equal?**

No. From the previous answers, $P(A=a1|B=b1) \approx 0.698$ while $P(B=b1|A=a1) \approx 0.617$, so they are not equal.

The mistaken belief that they are always equal is called the confusion of the inverse (also known as the conditional probability fallacy or the prosecutor's fallacy).

**8\. Does $P(A) = P(A|B)$? What does that mean about the independence of $A$ and $B$?**

No. $P(A=a1) = 0.60$ but $P(A=a1|B=b1) \approx 0.698$, so they are not equal.

Since conditioning on $B$ changes the probability of $A$, the variables $A$ and $B$ are **not independent** (they are dependent).

**9\. Using $P(A)$, $P(B|A)$, $P(B)$ from above, calculate,**

$P(A|B) = \frac{P(B|A)P(A)}{P(B)}$

Does it match your previous calculation for $P(A|B)$?

Using Bayes' theorem:

$P(A=a1|B=b1) = \dfrac{P(B=b1|A=a1)\,P(A=a1)}{P(B=b1)} = \dfrac{0.617 \times 0.60}{0.53} \approx \dfrac{0.37}{0.53} \approx 0.698$

$P(A=a2|B=b1) = \dfrac{0.400 \times 0.40}{0.53} \approx \dfrac{0.16}{0.53} \approx 0.302$

$P(A=a1|B=b2) = \dfrac{0.383 \times 0.60}{0.47} \approx \dfrac{0.23}{0.47} \approx 0.489$

$P(A=a2|B=b2) = \dfrac{0.600 \times 0.40}{0.47} \approx \dfrac{0.24}{0.47} \approx 0.511$

Yes, it matches the previous calculation for $P(A|B)$.

**10\. If we let A = H (some condition, characteristic, hypothesis) and B = D (some data, evidence, a test result), then how do we interpret each of the following: $P(H)$, $P(D)$, $P(H|D)$, $P(D|H)$, $P(H, D)$?**

$P(H)$ — the **prior**: the probability of the hypothesis/condition before observing any data.

$P(D)$ — the **marginal likelihood** (or evidence): the overall probability of seeing the data, across all hypotheses.

$P(H|D)$ — the **posterior**: the updated probability of the hypothesis after observing the data.

$P(D|H)$ — the **likelihood**: the probability of observing the data given that the hypothesis is true.

$P(H, D)$ — the **joint probability**: the probability of the hypothesis and the data occurring together.

## Bayes Rule

Bayes Rule will be an important part of our toolset in the weeks to come, especially in terms of Bayesian Inference. Work through the following problems in Bayes Rule.

### Problem 1 (Regular)

You might be interested in the relationship between alcoholism and liver disease, in which case “being an alcoholic” (or not) is a test (evidence for) for liver disease (or not).

Let `D` be the presence or absence of liver disease (`d` they have it; `~d`, "not d", they don't). Past data tells you that 10% of patients entering your clinic have liver disease. Let `A` be alcoholic (`a`) or not alcoholic (`~a`). 5% of the clinic’s patients are alcoholics.

You know that among those patients diagnosed with liver disease, 7% are alcoholics and among those without liver disease, 95.2% are non-alcoholics.

1. What is Bayes Rule for this problem? (write it out symbolically)
2. From the above word problem, what values of Bayes Rule do you have? Which ones are missing?
3. Calculate the missing values.
4. Calculate the posterior probability *distributions* using Bayes Rule.
5. Describe what each individual posterior probability means *in words*.

**1. Bayes Rule for this problem:**

$$P(D|A) = \frac{P(A|D)\,P(D)}{P(A)}$$

**2. Values we have and values that are missing:**

Have:
- $P(d) = 0.10$, so $P(\sim d) = 0.90$
- $P(a) = 0.05$, so $P(\sim a) = 0.95$
- $P(a|d) = 0.07$ (7% of those with liver disease are alcoholics)
- $P(\sim a|\sim d) = 0.952$ (95.2% of those without liver disease are non-alcoholics)

Missing:
- $P(a|\sim d)$
- The posteriors $P(d|a)$ and $P(d|\sim a)$

**3. Calculate the missing values:**

- $P(a|\sim d) = 1 - P(\sim a|\sim d) = 1 - 0.952 = 0.048$

**4. Posterior probability distributions using Bayes Rule:**

$P(d|a) = \dfrac{P(a|d)\,P(d)}{P(a)} = \dfrac{0.07 \times 0.10}{0.05} = 0.14$

$P(\sim d|a) = 1 - 0.14 = 0.86$

For $P(d|\sim a)$, first find $P(\sim a) = 1 - 0.05 = 0.95$:

$P(d|\sim a) = \dfrac{P(\sim a|d)\,P(d)}{P(\sim a)} = \dfrac{0.93 \times 0.10}{0.95} \approx 0.0979$

$P(\sim d|\sim a) = 1 - 0.0979 \approx 0.9021$

**5. What each posterior means in words:**

- $P(d|a) = 0.14$: if a patient is an alcoholic, there is a 14% chance they have liver disease.
- $P(\sim d|a) = 0.86$: if a patient is an alcoholic, there is an 86% chance they do not have liver disease.
- $P(d|\sim a) \approx 0.098$: if a patient is not an alcoholic, there is about a 9.8% chance they have liver disease.
- $P(\sim d|\sim a) \approx 0.902$: if a patient is not an alcoholic, there is about a 90.2% chance they do not have liver disease.

### Problem 2 (Harder)

In a particular pain clinic, 10% of patients are prescribed narcotic pain killers. Overall, five percent of the clinic’s patients are addicted to narcotics (including pain killers and illegal substances). Out of all the people prescribed pain pills, 8% are addicts. What is the relationship between addiction and pain pill prescriptions?

1. What is Bayes Rule for this problem? (write it out symbolically)
2. From the above word problem, what values of Bayes Rule do you have? Which ones are missing?
3. Calculate the missing values.
4. Calculate the posterior probability *distributions* using Bayes Rule.
5. Describe what each individual posterior probability means *in words*.

(Note: this problem is structured slightly differently than usual. You will need to use Total Probability and the Axioms of Probability as well as solving simultaneous equations to get the answer).

Let `P` be prescribed pain pills (`p`) or not (`~p`), and `A` be addicted (`a`) or not (`~a`).

**1. Bayes Rule for this problem:**

$$P(A|P) = \frac{P(P|A)\,P(A)}{P(P)}$$

**2. Values we have and values that are missing:**

Have:
- $P(p) = 0.10$, so $P(\sim p) = 0.90$
- $P(a) = 0.05$, so $P(\sim a) = 0.95$
- $P(a|p) = 0.08$ (8% of those prescribed pain pills are addicts)

Missing:
- $P(p|a)$ and $P(p|\sim a)$ (the likelihoods)
- The posteriors $P(a|p)$ is given, but we need $P(a|\sim p)$ and the full distributions

**3. Calculate the missing values:**

Joint for prescribed and addicted:

$P(a, p) = P(a|p)\,P(p) = 0.08 \times 0.10 = 0.008$

Likelihood $P(p|a)$:

$P(p|a) = \dfrac{P(a, p)}{P(a)} = \dfrac{0.008}{0.05} = 0.16$

Using Total Probability for the prescribed-but-not-addicted joint:

$P(\sim a, p) = P(p) - P(a, p) = 0.10 - 0.008 = 0.092$

Likelihood $P(p|\sim a)$:

$P(p|\sim a) = \dfrac{P(\sim a, p)}{P(\sim a)} = \dfrac{0.092}{0.95} \approx 0.0968$

**4. Posterior probability distributions using Bayes Rule:**

$P(a|p) = \dfrac{P(p|a)\,P(a)}{P(p)} = \dfrac{0.16 \times 0.05}{0.10} = 0.08$

$P(\sim a|p) = 1 - 0.08 = 0.92$

For the not-prescribed group, use $P(\sim p) = 0.90$:

$P(a, \sim p) = P(a) - P(a, p) = 0.05 - 0.008 = 0.042$

$P(a|\sim p) = \dfrac{P(a, \sim p)}{P(\sim p)} = \dfrac{0.042}{0.90} \approx 0.0467$

$P(\sim a|\sim p) = 1 - 0.0467 \approx 0.9533$

**5. What each posterior means in words:**

- $P(a|p) = 0.08$: if a patient is prescribed pain pills, there is an 8% chance they are addicted.
- $P(\sim a|p) = 0.92$: if a patient is prescribed pain pills, there is a 92% chance they are not addicted.
- $P(a|\sim p) \approx 0.047$: if a patient is not prescribed pain pills, there is about a 4.7% chance they are addicted.
- $P(\sim a|\sim p) \approx 0.953$: if a patient is not prescribed pain pills, there is about a 95.3% chance they are not addicted.

## Titanic

In [3]:
import pandas as pd
from pandasql import sqldf
import seaborn as sns
import matplotlib.pyplot as plt

Make sure you worked through the Titanic case study. This is a continuation of that analysis. *Feel free to copy code blocks from the case study as you see fit*

We start by loading the data:

In [5]:
titanic = pd.read_csv("https://raw.githubusercontent.com/fundamentals-of-data-science/datasets/master/titanic.csv")
titanic.head()

,pclass,survived,name,sex,age,sibsp,parch,ticket,fare,cabin,embarked,boat,body,home.dest
0,1.0,1.0,"Allen, Miss. Elisabeth Walton",female,29.0000,0.0,0.0,24160,211.3375,B5,S,2,NaN,"St Louis, MO"
1,1.0,1.0,"Allison, Master. Hudson Trevor",male,0.9167,1.0,2.0,113781,151.5500,C22 C26,S,11,NaN,"Montreal, PQ / Chesterville, ON"
2,1.0,0.0,"Allison, Miss. Helen Loraine",female,2.0000,1.0,2.0,113781,151.5500,C22 C26,S,NaN,NaN,"Montreal, PQ / Chesterville, ON"
3,1.0,0.0,"Allison, Mr. Hudson Joshua Creighton",male,30.0000,1.0,2.0,113781,151.5500,C22 C26,S,NaN,135.0,"Montreal, PQ / Chesterville, ON"
4,1.0,0.0,"Allison, Mrs. Hudson J C (Bessie Waldo Daniels)",female,25.0000,1.0,2.0,113781,151.5500,C22 C26,S,NaN,NaN,"Montreal, PQ / Chesterville, ON"


In [6]:
display(titanic.info())

<class 'pandas.DataFrame'>
RangeIndex: 1310 entries, 0 to 1309
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   float64
 1   survived   1309 non-null   float64
 2   name       1309 non-null   str    
 3   sex        1309 non-null   str    
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   float64
 6   parch      1309 non-null   float64
 7   ticket     1309 non-null   str    
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    str    
 10  embarked   1307 non-null   str    
 11  boat       486 non-null    str    
 12  body       121 non-null    float64
 13  home.dest  745 non-null    str    
dtypes: float64(7), str(7)
memory usage: 143.4 KB


None

## Conditional Probabilities

#### __1. Calculate $P(survived|parch)$__

**Discussion:** P(survived∣parch) is the probability that a passenger survived *given* how many parents or children they had aboard. Each row is a value of `parch` (the given) and the columns are the survival outcome, so rows sum to 1. Since `parch` mixes "has a parent aboard" with "has children aboard," I expect a non-monotonic pattern. My hypothesis: passengers with 0 family members (the large single-traveler group) survive at a lower rate, while those with a small number of parents/children (1 to 3) do better, possibly because of the "women and children first" policy and families being prioritized. I'd expect the very high counts (4+) to be noisy because the cell sizes are tiny.

In [7]:
P_survived_given_parch = pd.crosstab(
    titanic["parch"], titanic["survived"], normalize="index"
)
P_survived_given_parch

survived,0.0,1.0
parch,,
0.0,0.664671,0.335329
1.0,0.411765,0.588235
2.0,0.495575,0.504425
3.0,0.375000,0.625000
4.0,0.833333,0.166667
5.0,0.833333,0.166667
6.0,1.000000,0.000000
9.0,1.000000,0.000000


**Discuss (results vs. hypothesis):** The pattern matches the hypothesis at the low end. Solo travelers (parch=0) had only a 33.5% survival rate, the worst of any well-populated group, consistent with most of them being single third-class men. Passengers with 1 to 3 parents/children survived at much higher rates (58.8%, 50.4%, 62.5%), supporting the idea that being part of a small family helped, likely through the "women and children first" dynamic.
The tail (parch = 4, 5, 6, 9) shows low survival, but these cells are tiny (only 2 to 6 people each), so those estimates are unreliable and should not be over-interpreted. There may also be a real effect that large families, often poorer and in third class, struggled to evacuate together. The takeaway: parch is associated with survival, but the relationship is non-monotonic and the extreme values are too sparse to draw conclusions from.

#### __2. Calculate $P(survived|fare)$__

**Discussion:** P(survived∣fare) is the probability a passenger survived *given* the fare they paid. Fare is continuous and strongly right-skewed (a power law, mean 33 GBP but max 512 GBP), so a raw conditional on every distinct fare would give one-person cells and meaningless probabilities. Following the notebook's approach with `age` to `decade`, I discretize fare into 50-GBP buckets so each value of the "given" has enough evidence. My hypothesis: fare is a proxy for passenger class and wealth, and since first-class passengers survived at higher rates, I expect survival probability to rise as fare increases. I also expect the highest buckets to be sparse and therefore noisy.

In [8]:
titanic["fare_bucket"] = (titanic["fare"] // 50) * 50

P_survived_given_fare = pd.crosstab(
    titanic["fare_bucket"], titanic["survived"], normalize="index"
)
P_survived_given_fare

survived,0.0,1.0
fare_bucket,,
0.0,0.679174,0.320826
50.0,0.379747,0.620253
100.0,0.212121,0.787879
150.0,0.384615,0.615385
200.0,0.428571,0.571429
250.0,0.230769,0.769231
500.0,0.000000,1.000000


**Discuss (results vs. hypothesis):** The result supports the hypothesis at the coarse level. The cheapest bucket (0–49 GBP), which holds the overwhelming majority of passengers (1066 of them, mostly third class), had only a 32% survival rate. Every higher fare bucket flips to majority-survived, with the 50–99 and 100–149 bands jumping to roughly 62% and 79%. So paying more was clearly associated with surviving, consistent with fare acting as a proxy for class and access to lifeboats.
However, the relationship is not cleanly monotonic above 50 GBP. The 150–199 and 200–249 buckets dip back toward 57–62% rather than continuing to climb. This is almost certainly a sample-size artifact: those buckets contain only 13 and 21 people, so a few individuals swing the percentage. The 500+ bucket showing 100% survival rests on just 4 passengers and should not be believed as a model, exactly the caution the notebook raises with the "all 80-year-olds survive" example. The summary: fare is positively associated with survival, the signal is strong at the low end where the data is dense, and the wiggle at the top is noise from thin cells rather than a real reversal.

## Naive Bayes Classifier

#### __1. Calculate the Naive Bayes Classifier for $P(survived|pclass, sex, decade, parch, sibsp)$ and make 5 predictions.__

**Discussion:** This is the full Naive Bayes Classifier the lab calls for, extending the notebook's two-feature (pclass, sex) model to five features. The "naive" assumption is that the features are conditionally independent given survival, so the posterior factors as P(survived∣x)∝P(survived)∏iP(xi∣survived)P(survived|\mathbf{x}) \propto P(survived)\prod_i P(x_i|survived)
P(survived∣x)∝P(survived)∏i​P(xi​∣survived). That assumption is known to be violated here (pclass, fare, family variables are correlated), but the classifier often still ranks outcomes correctly, which is what we care about. I follow the notebook by deriving decade from age and restricting to passengers with a known age so decade is defined. From the earlier EDA, I expect sex and pclass to dominate: women and high-class passengers should be predicted to survive, solo third-class men to perish.

In [14]:
from sklearn.naive_bayes import CategoricalNB
from sklearn.preprocessing import OrdinalEncoder

import warnings
warnings.filterwarnings("ignore")

titanic["decade"] = (titanic["age"] // 10) * 10

features = ["pclass", "sex", "decade", "parch", "sibsp"]
data = titanic.dropna(subset=features + ["survived"]).copy()   # 1046 rows with known age

encoder = OrdinalEncoder()
encoder.fit(data[features])

clf = CategoricalNB()
clf.fit(encoder.transform(data[features]), data["survived"])

# 5 made-up passengers: (pclass, sex, decade, parch, sibsp)
passengers = [
    (1, "female", 30, 0, 1),   # wealthy woman, husband aboard
    (3, "male",   20, 0, 0),   # young solo immigrant man
    (2, "female", 30, 2, 1),   # mid-class mother, 2 kids + spouse
    (1, "male",   50, 0, 1),   # older wealthy man with wife
    (3, "male",    0, 2, 1),   # young boy, 3rd class, with family
]

for p in passengers:
    enc = encoder.transform([p])
    proba = clf.predict_proba(enc)[0]
    pred = "SURVIVE" if clf.predict(enc)[0] == 1 else "PERISH"
    print(p, f"P(perish)={proba[0]:.3f}  P(survive)={proba[1]:.3f}  -> {pred}")

(1, 'female', 30, 0, 1) P(perish)=0.090  P(survive)=0.910  -> SURVIVE
(3, 'male', 20, 0, 0) P(perish)=0.930  P(survive)=0.070  -> PERISH
(2, 'female', 30, 2, 1) P(perish)=0.084  P(survive)=0.916  -> SURVIVE
(1, 'male', 50, 0, 1) P(perish)=0.501  P(survive)=0.499  -> PERISH
(3, 'male', 0, 2, 1) P(perish)=0.530  P(survive)=0.470  -> PERISH


**Discuss (results vs. expectation):** The first three predictions land exactly where the EDA pointed. The wealthy first-class woman and the second-class mother are both predicted to survive with ~91% confidence, while the young solo third-class man is predicted to perish at 93%, all consistent with "women and children first" plus a strong class effect.

The two interesting cases are where the features conflict. Passenger 4, a 50s first-class man, comes out an almost exact coin flip (0.499). His class pushes toward survival but his sex pushes hard the other way, and the naive model splits the difference, which matches my own uncertainty. Passenger 5 is the one that contradicts my prior: I expected a small child to be saved, but the model predicts perish (47% survive). This is the Naive Bayes assumption biting. The boy's third-class status and his presence in a larger family (parch=2, sibsp=1) each independently drag the probability down, and because the model treats those signals as independent rather than recognizing "young child traveling with family," it does not give childhood the protective weight the real evacuation did. It is a good illustration that the conditional-independence assumption can misfire exactly when features interact, even when the classifier gets the easy, sex-and-class-dominated cases right.